src/app/admin/trace_level.rs

```diff
@@ -0,0 +1,85 @@
+use super::rsp;
+use control::ClientAddr;
+use futures::{
+    future::{self, Future},
+    Stream,
+};
+use http::{Method, StatusCode};
+use hyper::{service::Service, Body, Request, Response};
+pub use trace::LevelHandle as TraceLevel;
+
+use std::{io, net::SocketAddr, str};
+
+impl Service for TraceLevel {
+    type ReqBody = Body;
+    type ResBody = Body;
+    type Error = io::Error;
+    type Future = Box<Future<Item = Response<Body>, Error = Self::Error> + Send + 'static>;
+
+    fn call(&mut self, req: Request<Body>) -> Self::Future {
+        // `/proxy-log-level` endpoint can only be called from loopback IPs
+        if let Some(addr) = req.extensions().get::<ClientAddr>() {
+            let addr: SocketAddr = addr.into();
+            if !addr.ip().is_loopback() {
+                warn!(message = "denying request from non-loopback IP", %addr);
+                return Box::new(future::ok(rsp(
+                    StatusCode::FORBIDDEN,
+                    "access to /proxy-log-level only allowed from loopback interface",
+                )));
+            }
+        } else {
+            // TODO: should we panic if this was unset? It's a bug, but should
+            // it crash the proxy?
+            error!(message = "ClientAddr extension should always be set");
+            return Box::new(future::ok(rsp(
+                StatusCode::INTERNAL_SERVER_ERROR,
+                Body::empty(),
+            )));
+        }
+
+        match req.method() {
+            &Method::GET => match self.current() {
+                Ok(level) => Box::new(future::ok(rsp(StatusCode::OK, level))),
+                Err(error) => {
+                    warn!(message = "error getting proxy log level", %error);
```

Comment:
What does the % symbol mean here?

Reply:
This is using the `tracing` crate's [structured key-value fields](https://docs.rs/tracing/0.1.2/tracing/macro.span.html#recording-fields); the `%` symbol means that the field should be recorded using its `fmt::Display` implementation.